### PR TITLE
[idrac_system_erase] - Erase of system and storage components

### PR DESCRIPTION
**Pull request by rajshekarp87**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### OUTPUT

```paste below

```
##### ADDITIONAL INFORMATION

```paste below

```
# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

### CODE REVIEW

**Comment by ansible-collections-svc** [comment]
Can one of the admins verify this patch?

File: docs/modules/idrac_system_erase.rst [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+.. contents::
+   :local:
+   :depth: 1
+
+
+Synopsis
+--------
+
+This module allows erasing System components such as iDRAC, BIOS, DIAG, etc., and Storage components such as PERC NV cache, Non-volatile memory, Cryptographic erase PD, etc. of the server.
```
Comment (reetivel) [inline docs/modules/idrac_system_erase.rst]: This module allows you to erase system components such as iDRAC, BIOS, DIAG, and so forth. You can also erase storage components such as PERC NV cache, non-volatile memory, cryptographic erase of physical disks, and so on of the server.

File: docs/modules/idrac_system_erase.rst [hunk 2]
```diff
@@ -0,0 +1,188 @@
+.. _idrac_system_erase_module:
+
+
+idrac_system_erase -- Erase system and storage components of the server
```
Comment (reetivel) [inline docs/modules/idrac_system_erase.rst:4]: Erase system components and storage components of the server.

File: docs/modules/idrac_system_erase.rst [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+Synopsis
+--------
+
+This module allows erasing System components such as iDRAC, BIOS, DIAG, etc., and Storage components such as PERC NV cache, Non-volatile memory, Cryptographic erase PD, etc. of the server.
+
+
+
+Requirements
+------------
+The below requirements are needed on the host that executes this module.
```
Comment (reetivel) [inline docs/modules/idrac_system_erase.rst:21]: The following requirements are a must to run this module on the host.

File: docs/modules/idrac_system_erase.rst [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+- python \>= 3.9.6
+
+
+
+Parameters
+----------
+
+  component (True, list, None)
+    System and storage components to erase.
```
Comment (reetivel) [inline docs/modules/idrac_system_erase.rst]: List of system and storage components that can be deleted.
The following are the supported components:
- AllApps 
- BIOS
- CryptographicErasePD 
- DIAG 
- DPU 
- DrvPack 
- IDRAC
- LCData
- NonVolatileMemory 
- OverwritePD 
- PERCNVCache
- ReinstallFW 
- vFlash

File: docs/modules/idrac_system_erase.rst [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    Below are the supported components. AllApps BIOS CryptographicErasePD DIAG DPU DrvPack IDRAC LCData NonVolatileMemory OverwritePD PERCNVCache ReinstallFW vFlash
```
Comment (reetivel) [inline docs/modules/idrac_system_erase.rst]: Added in above comment

File: plugins/modules/idrac_system_erase.py [hunk 6]
```diff
@@ -0,0 +1,569 @@
...
+        OverwritePD
+        PERCNVCache
+        ReinstallFW
+        vFlash
+    type: list
+    elements: str
+    required: true
+  power_on:
+    description:
+      - This allows to power on the server after erase operation. This is applicable only when I(job_wait) is C(true).
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py]: This parameter allows you to power on the server after the erase operation is completed. This is applicable when I(job_wait) is C(true).

File: plugins/modules/idrac_system_erase.py [hunk 7]
```diff
@@ -0,0 +1,569 @@
...
+      - C(true) will power on the server.
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py]: - C(true) power on the server.
 - C(false) does not power on the server.

File: plugins/modules/idrac_system_erase.py [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+      - C(false) will not power on the server.
+    type: bool
+    default: false
+  job_wait:
+    description:
+      - Whether to wait till completion of the job. This is applicable when I(power_on) is C(true).
+      - C(true) will wait for job completion.
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py]: - C(true) waits for the job completion.
- C(false) does not wait for the job completion.

File: plugins/modules/idrac_system_erase.py [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+    description:
+      - Manager ID of the iDRAC.
+    type: str
+requirements:
+  - "python >= 3.9.6"
+author:
+  - "Rajshekar P(@rajshekarp87)"
+attributes:
+    check_mode:
+        description: Runs task to validate without performing action on the target machine.
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py:79]: Runs task to validate (validate what?... please specify) without acting on the target machine.
 or
  Runs task to validate (validate what?... please specify) without performing (action/operation name) operation on the target machine.

File: plugins/modules/idrac_system_erase.py [hunk 10]
```diff
@@ -0,0 +1,569 @@
...
+        support: full
+    diff_mode:
+        description: Runs the task to report the changes made or to be made.
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py:82]: Runs the task to report the changes applied or to be applied.

File: plugins/modules/idrac_system_erase.py [hunk 11]
```diff
@@ -0,0 +1,569 @@
...
+        support: none
+notes:
+    - Run this module from a system that has direct access to Dell iDRAC.
+    - This module supports only iDRAC9 and above.
+    - This module supports IPv4 and IPv6 addresses.
+"""
+
+EXAMPLES = r"""
+---
+- name: Erase single component and power on the server
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py]: - name: Erase a single component and power on the server.

File: plugins/modules/idrac_system_erase.py [hunk 12]
```diff
@@ -0,0 +1,569 @@
...
+  dellemc.openmanage.idrac_system_erase:
+    idrac_ip: 198.162.0.1
+    idrac_user: username
+    idrac_password: passw0rd
+    ca_path: "/path/to/ca_cert.pem"
+    component: ["BIOS"]
+    power_on: true
+
+- name: Erase multiple components and don't power on the server after erase operation
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py]: - name: Erase multiple servers or storage components and do not power on the server after the erase operation is completed.

File: plugins/modules/idrac_system_erase.py [hunk 13]
```diff
@@ -0,0 +1,569 @@
...
+  dellemc.openmanage.idrac_system_erase:
+    idrac_ip: 198.162.0.1
+    idrac_user: username
+    idrac_password: passw0rd
+    ca_path: "/path/to/ca_cert.pem"
+    component: ["BIOS", "DIAG", "PERCNVCache"]
+
+- name: Erase multiple components and don't wait for the job completion
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py]: - name: Erase multiple storage and server components and do not wait for the job completion.

File: plugins/modules/idrac_system_erase.py [hunk 14]
```diff
@@ -0,0 +1,569 @@
...
+    idrac_password: passw0rd
+    ca_path: "/path/to/ca_cert.pem"
+    component: ["IDRAC", "DPU", "LCData"]
+    job_wait: false
+"""
+
+RETURN = r'''
+---
+msg:
+    description: Status of the component system erase operation.
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py:122]: description: Status of the component system erase operation.

File: plugins/modules/idrac_system_erase.py [hunk 15]
```diff
@@ -0,0 +1,569 @@
...
+    returned: always
+    type: str
+    sample: "Successfully completed the system erase operation."
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py:125]: sample: "Successfully completed the system erase operation."

File: plugins/modules/idrac_system_erase.py [hunk 16]
```diff
@@ -0,0 +1,569 @@
...
+job_details:
+    description: Returns the output for status of the job.
+    returned: For system erase operation
```
Comment (reetivel) [inline plugins/modules/idrac_system_erase.py:128]: For system erase operation

**Review by reetivel — COMMENTED** [review]
- Added editorial review comments.

- The term "Erase" is not supported as per Dell's Guidelines (Acrolinx terminology).
Please use any of the following terms wherever applicable:
a. "Delete" - preferred
b. "Remove"

File: plugins/modules/idrac_system_erase.py [hunk 17]
```diff
@@ -0,0 +1,569 @@
...
+        if res_id:
+            uri = MANAGERS_URI + "/" + res_id
+        else:
+            uri, error_msg = validate_and_get_first_resource_id_uri(
+                self.module, self.idrac, MANAGERS_URI)
+            if error_msg:
+                self.module.exit_json(msg=error_msg, failed=True)
+        return uri
+
+    def get_job_status(self, erase_component_response):
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_system_erase.py:291]: Can you please check if we have same function in utils if not please move it.

File: plugins/modules/idrac_system_erase.py [hunk 18]
```diff
@@ -0,0 +1,569 @@
...
+            self.module.exit_json(msg=WAIT_TIMEOUT_MSG.format(
+                job_wait_timeout), changed=True, job_status=job_dict)
+        if job_failed:
+            self.module.exit_json(
+                msg=job_dict.get('Message'),
+                failed=True,
+                job_details=job_dict)
+        return job_dict
+
+    def get_job_details(self, erase_component_response):
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_system_erase.py:323]: Can you please check if we have same function in utils if not please move it.

File: docs/modules/idrac_system_erase.rst (hunk 4 above)
Comment (rajshekarp87) [reply docs/modules/idrac_system_erase.rst]: Done

File: plugins/module_utils/utils.py [hunk 19]
```diff
@@ -569,6 +569,17 @@ def trigger_restart_operation(idrac, restart_type="GracefulRestart", resource_id
     return resp, error_msg
 
 
+def power_on_operation(idrac, restart_type="On", resource_id=None):
```
Comment (ABHISHEK-SINHA10) [inline plugins/module_utils/utils.py]: There's already function to restart the idrac, we can utilize the same. This function is duplicate of trigger_restart_operation.

File: docs/modules/idrac_system_erase.rst (hunk 1 above)
Comment (rajshekarp87) [reply docs/modules/idrac_system_erase.rst:15]: Done

File: docs/modules/idrac_system_erase.rst (hunk 2 above)
Comment (rajshekarp87) [reply docs/modules/idrac_system_erase.rst:4]: Done

File: docs/modules/idrac_system_erase.rst (hunk 5 above)
Comment (rajshekarp87) [reply docs/modules/idrac_system_erase.rst]: Done

File: plugins/module_utils/utils.py (hunk 19 above)
Comment (rajshekarp87) [reply plugins/module_utils/utils.py]: Done

File: docs/modules/idrac_system_erase.rst (hunk 1 above)
Comment (rajshekarp87) [reply docs/modules/idrac_system_erase.rst]: This is generated by ansible-doc-extractor. Kindly ignore reviewing RST file moving forward
RST content is updated as per the module

File: plugins/modules/idrac_system_erase.py (hunk 17 above)
Comment (rajshekarp87) [reply plugins/modules/idrac_system_erase.py:291]: Its not available in the utils. Keeping this in the module itself. Will take care of adding to utils in the upcoming release

File: plugins/modules/idrac_system_erase.py (hunk 18 above)
Comment (rajshekarp87) [reply plugins/modules/idrac_system_erase.py:323]: Its not available in the utils. Keeping this in the module itself. Will take care of adding to utils in the upcoming release

**Review by ABHISHEK-SINHA10 — DISMISSED** [review]

**Comment by ShivamSh3** [comment]
image

**Review by sachin-apa — DISMISSED** [review]

**Review by sachin-apa — APPROVED** [review]

**Review by ShivamSh3 — APPROVED** [review]

**Pull request merged**